### ISSUE TITLE
Abstract class implementations don't throw an error when an optional parameter is not kept optional 

### ISSUE THREAD

**Issue by diesal11** [microsoft/TypeScript#38019]
**TypeScript Version:**  nightly (3.9.0)

**Search Terms:**
Abstract class optional undefined parameter implementation incorrect type 

**Code**

```ts
abstract class A {
    abstract foo(bar?: string): void
}

class B implements A {
    foo(bar: string): void {
        bar.toLowerCase()
    }
}

const obj: A = new B()
obj.foo() // bar will be undefined and .toLowerCase will cause a runtime error
```

**Expected behavior:**
The compiler should warn or error that B is not a valid implementation of A as the argument types for foo don't match (string vs string | undefined).

**Actual behavior:**
The compiler throws no such warning and the resulting code produces a runtime error when run

**Playground Link:** [here](https://www.typescriptlang.org/play/?ts=3.9.0-dev.20200416&ssl=1&ssc=1&pln=12&pc=79#code/IYIwzgLgTsDGEAJYBthjAgggg3gKAUIVEhngQDMB7KgChGCgH4AuBUgSwDsBzASjYA3KhwAmeAL548KNBgBCCDgFsADsgCmyjVwgZs+IpRr1GbTrwEJhY3ASNEGUAHQQqAGSoB3DVADCaBq0fPaEUlIyVFyQCFQgAFZs2AC8CFwaXgjywXhx8c7UdHwIAPQlCE4IXhzIyBUaCACuXKIaFNwaosQtCK4e3r4BYA3VtUjAjcPECFDNECoNvlBUUEA)

**Related Issues:** None that i could find

**Comment by RyanCavanaugh** [comment]
Method signatures are compared bivariantly because it's common for classes to vary in this way

**Comment by MicahZoltu** [comment]
@RyanCavanaugh Where can I read more about why this is by design?  This feels very much like a bug to me, and your explanation unfortunately didn't really help clear it up.  The example given in the OP shows how this behavior can easily turn into a runtime failure, and the fix feels "obvious" to me.

Note for future readers: This isn't just an issue with `undefined` or classes or abstract classes.
```ts
interface Fruit {
    eat(options: string | number): void
}
interface Apple extends Fruit {
    eat(options: string): void
}
declare const apple: Apple
const fruit: Fruit = apple
apple.eat(5) // error (correct)
fruit.eat(5) // ok (incorrect IMO)
```

**Comment by MicahZoltu** [comment]
I found the answer (I haven't read it yet though): https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant
